### PR TITLE
Feature/contact refactoring

### PR DESCRIPTION
**Pull request by janherich**
Refactoring of contacts/contact-groups and associated events, prerequisite for re-enabling group chats:
1. No more `group-contact`/`chat-contact` objects, as it's just a simple array of ids pointing to `contact` objects in realm
2. No more dispatching like crazy just to initialise app-db with correct persisted/default contacts/contact-groups
3. Removal of redundant information in `default_contacts.json`/`default_contact_groups.json` -> previously each contact groups specified list of DApps and at the same time, each DApp specified list of contacts groups where it belonged. We somehow reflected both in the code, overwriting first with the second, which makes no sense as there should be just one place to specify the relation. I Decided that it's more succinct and easier (for the implementation) to have this info in contact group, if we decide the the opposite is better, I can easily change it.

### Notes for reviewers: 
Don't be put off by the number of changes, it's mostly `package-lock.json` (necessary as realm version was updated).

### Notes for QA: 
Basic regression testing is necessary, making sure that commands, changing accounts, contacts (persistance, etc.) are not broken.
I also re-enabled private group chats in the PR, along with adding/removing users from conversation.
There are still couple of quirks which I need to solve (system messages behaving odd sometimes), but those will be solved in another PRs.
The point re group chats is that they shouldn't produce any errors like before and should be basically usable (admin is able to moderate the group and chat with people).

### CODE REVIEW

**Review by yenda — APPROVED** [review]

**Review by cammellos — APPROVED** [review]

**Comment by rasom** [comment]
@janherich are there changes which rely on a features from a newer version of realm?

**Comment by janherich** [comment]
@rasom yes, they are (being able to hold list of strings as prop, instead of a list of child objects)

File: src/status_im/chat/events.cljs [hunk 1]
```diff
@@ -81,36 +84,56 @@
...
+  (when-not (get-in db [:chats constants/console-chat-id])
+    {:db                   (-> db
+                               (assoc :current-chat-id constants/console-chat-id)
+                               (update :chats assoc constants/console-chat-id console/chat))
+     :data-store/save-chat console/chat}))
+
+(defn- add-default-contacts
+  [{:keys [db default-contacts] :as cofx}]
+  (let [new-contacts      (-> {}
+                              (into (map (fn [[id props]]
```
Comment (rasom) [inline src/status_im/chat/events.cljs:98]: Can we extract function here?

File: src/status_im/chat/events.cljs [hunk 2]
```diff
@@ -81,36 +84,56 @@
...
+                                                          :hide-contact?    (-> props :hide-contact? boolean)
+                                                          :pending?         (-> props :pending? boolean)
+                                                          :dapp?            (:dapp? props)
+                                                          :dapp-url         (-> props :dapp-url :en)
+                                                          :bot-url          (:bot-url props)
+                                                          :description      (:description props)}])))
+                                    default-contacts)
+                              (assoc constants/console-chat-id console/contact))
+        existing-contacts (:contacts/contacts db)
+        contacts-to-add   (select-keys new-contacts (set/difference (set (keys new-contacts)) (set (keys existing-contacts))))]
```
Comment (rasom) [inline src/status_im/chat/events.cljs]: Please keep lines less than 80 chars

File: src/status_im/chat/events.cljs [hunk 3]
```diff
@@ -126,11 +149,13 @@
...
-      (-> db
-          (assoc :chats chats
-                 :deleted-chats inactive-chat-ids)
-          init-console-chat
-          (update :dispatch-n conj [:load-default-contacts!])))))
+      (handlers/merge-fx cofx
+                         {:db (assoc db
+                                     :chats chats
+                                     :deleted-chats inactive-chat-ids)}
+                         (init-console-chat)
```
Comment (rasom) [inline src/status_im/chat/events.cljs:158]: Can we skip using extra parentheses here?

**Review by rasom — APPROVED** [review]

File: src/status_im/chat/events.cljs (hunk 3 above)
Comment (janherich) [reply src/status_im/chat/events.cljs:158]: Unfortunately, that would require changing the `merge-fx` macro (so it works the same as core `->` and `->>` macros where no parenthesis is required when s-expression is to be evaluated with only one argument coming from the macro itself).
We can do it, but I would prefer to do it in separate PR.

**Review by goranjovic — APPROVED** [review]

File: src/status_im/chat/events.cljs (hunk 3 above)
Comment (rasom) [reply src/status_im/chat/events.cljs:158]: Oh ok, sure. Idk if it makes sense but we could rename it to `merge-fx->`, definitely not for this PR.

File: src/status_im/transport/message/v1/protocol.cljs [hunk 4]
```diff
@@ -63,11 +63,17 @@
...
+            :sym-key-id sym-key-id}
+           {:topic   (transport.utils/get-topic constants/contact-discovery)
+            :pub-key public-key}))
        public-keys))
 
 (defn multi-send-with-pubkey
-  "Sends payload to multiple participants selected by `:public-keys` key."
+  "Sends payload to multiple participants selected by `:public-keys` key.
+  If there is already established symmetric key with the participant, uses that (for efficiency),
+  if not, uses asymetric encryption."
```
Comment (yenda) [inline src/status_im/transport/message/v1/protocol.cljs:76]: since the admin only adds his contacts to the group there should always be an established symmetric key why is this change needed ?

Comment (janherich) [reply src/status_im/transport/message/v1/protocol.cljs:76]: 1. We are using this method to send contact updates as well (potentially to people who you have in contacts but you have no open chats with them, so no symmetric key)
2. `multi-send-with-pubkey` is used to actually send group key to all the participants (`NewGroupKey`), for example when new group chat is first established, or new person is added - we must use asymmetric encryption there (barring the "lucky" case we can re-use topic + symkey of already established chat)

Comment (yenda) [reply src/status_im/transport/message/v1/protocol.cljs:76]: Ok makes sense for contact updates, but for NewGroupKey I don't see why we can't send with symmetric encryption since the admin is a contact of all members. We should relabel the function multi-send since it's not only sending with pubkey anymore though

Comment (janherich) [reply src/status_im/transport/message/v1/protocol.cljs:76]: I don't want to make assumption that when I'm somebody's contact, I have always chat with negotiated symmetric key opened with him - for example once I delete the chat, I don't, even when we are still contacts. 
I agree about renaming the function, plain `multi-send` is much less confusing.

Comment (yenda) [reply src/status_im/transport/message/v1/protocol.cljs:76]: If you don't want to make that assumption then you should always send with public key ?

Comment (janherich) [reply src/status_im/transport/message/v1/protocol.cljs:76]: That's probably cleaner code wise, but using the existing symmetric key established for (potentially) existing 1-1 chat with particular contact is just a minor perf optimisation.

**Comment by asemiankevich** [comment]
there are some issues with the PR like re-adding member and this `system` saying that someone is added or removed and dapps are in the list when i create group chat but we spoke with @janherich and this will be polished in next PRs
